iroh-rpc/src/stream.rs

```diff
@@ -1,96 +1,190 @@
+        let task = Arc::new(Mutex::new(AtomicWaker::new()));
+        OrderedStream {
             out_sender,
+            error_receiver,
             next_chunk: 0,
             id: header.id,
             num_chunks: header.num_chunks,
-            chunks: HashMap::new(),
+            chunks: Arc::clone(&chunks),
+            task: Arc::clone(&task),
+            handle: {
+                tokio::spawn(async move {
+                    OrderedStream::gather_packets(
+                        packet_receiver,
+                        chunks,
+                        task,
+                        error_sender,
+                        shutdown_receiver,
+                    )
+                    .await
+                })
+            },
+            early_shutdown: Some(shutdown_sender),
         }
     }
+
+    // we have loop that is taking in packets and ordering them. If there is an
+    // error, send the error to the main stream
+    async fn gather_packets(
+        mut packet_receiver: StreamReceiver,
+        chunks: Arc<Mutex<HashMap<u64, Vec<u8>>>>,
+        task: Arc<Mutex<AtomicWaker>>,
+        mut error_sender: mpsc::Sender<RpcError>,
+        mut shutdown: oneshot::Receiver<()>,
+    ) {
+        loop {
+            select! {
+              res = packet_receiver.next() => {
+                if let Some(r) = res {
+                  match r {
+                    StreamType::Packet(p) => {
+                      let mut chunks = chunks.lock().expect("failed to lock stream chunk mutex");
+                      chunks.insert(p.index, p.data);
+                      debug!(target="OrderedStream", "stream received chunk {}", p.index);
+                  },
+                    StreamType::RpcError(e) => {
+                      error!(target="OrderedStream", "stream received error {}", e);
+                      error_sender
+                          .send(e)
+                          .await
+                          .expect("failed to send on error sender");
+                    },
+                  }
+                  // notify poll_next that we have a message we want them to
+                  // process
+                  let task = task.lock().expect("failed to lock task");
+                  task.wake();
+                  continue;
+                }
+                // if res == None, then the packet_receiver stream has closed
+                    debug!(target="OrderedStream", "packet receiver closed, done
+                      gathering packets");
+                      return;
+                                },
+              _ = shutdown => {
+                debug!(target="OrderedStream", "early shutdown triggered, no longer
+                  gathering packets");
+                error_sender.send(RpcError::StreamClosedEarly).await.expect("failed to send on
+                  error_sender");
+                let task = task.lock().expect("failed to lock task");
```

Comment:
why do you need this task waker?

Reply:
Ah, this was from a previous refactor before we had the "just write an async `close` method to do clean up" conversation.

Previously, I had a method `shutdown` that sent on the `shutdown` channel & would cause this loop to break, but didn't do anything to stop someone from calling `next` on the stream again. This was my way of ensuring that if someone called `next` on the stream, you would get an error informing you the stream closed early. Now, the `close` method takes ownership of the `OrderedStream` struct, so you can't call `next` on it anyway. 

Thank you I will clean this up.

Reply:
sending to the channel will wakeup the task already, so I don't understand why you need this taskwaker in addition.

Reply:
I'm a bit confused. Are you taking about having the `AtomicWaker` in general, or just this particular line. This particular line, it's definitely useless.

But when I remove the `task` & don't call `task.wake()` at all in `gather_packets`, then `poll_next` gets called once and is `Pending` forever.

I could also just be doing something wrong or not leveraging some underlying rust things that I don't know about!

Reply:
to put a cap on this convo:
dig recommended using the `async_stream` `stream!` macro to get around having to futz with `poll_next` & `AtomicWaker`s